File: 3544.subtree-inversion-sum.java
#
# @lc app=leetcode id=3544 lang=java
#
# [3544] Subtree Inversion Sum
#
# @lc code=start
import java.util.*;
class Solution {
    public long subtreeInversionSum(int[][] edges, int[] nums, int k) {
        int n = nums.length;
        List<Integer>[] tree = new ArrayList[n];
        for (int i = 0; i < n; ++i) tree[i] = new ArrayList<>();
        for (int[] e : edges) {
            tree[e[0]].add(e[1]);
            tree[e[1]].add(e[0]);
        }
        // dp[u][d]: max sum for subtree rooted at u, d is distance to last inverted ancestor
        Long[][] dp = new Long[n][k+2];
        return dfs(0, -1, k, tree, nums, k, dp);
    }
    private long dfs(int u, int parent, int dist, List<Integer>[] tree, int[] nums, int k, Long[][] dp) {
        if (dp[u][dist] != null) return dp[u][dist];
        // Case 1: do not invert at u
        long sumNoInvert = nums[u];
        for (int v : tree[u]) {
            if (v == parent) continue;
            sumNoInvert += dfs(v, u, Math.min(dist+1, k+1), tree, nums, k, dp);
        }
        long res = sumNoInvert;
        // Case 2: invert at u (only if dist >= k)
        if (dist >= k) {
            long sumInvert = -nums[u];
            for (int v : tree[u]) {
                if (v == parent) continue;
                sumInvert += dfs(v, u, 1, tree, nums, k, dp);
            }
            res = Math.max(res, sumInvert);
        }
        return dp[u][dist] = res;
    }
}
# @lc code=end

File: 3544.subtree-inversion-sum.go
#
# @lc app=leetcode id=3544 lang=golang
#
# [3544] Subtree Inversion Sum
#

# @lc code=start
func subtreeInversionSum(edges [][]int, nums []int, k int) int64 {
    // Step 1: Convert edges into an adjacency list for easy tree traversal.
    adjList := make(map[int][]int)
    for _, edge := range edges {
        u, v := edge[0], edge[1]
        adjList[u] = append(adjList[u], v)
        adjList[v] = append(adjList[v], u)
    }

    // Step 2: Implement DFS with dynamic programming to calculate potential sums.
    var dfs func(node, parent int) (sumWithoutInversion, sumWithInversion int64)
    dfs = func(node, parent int) (sumWithoutInversion, sumWithInversion int64) {
        currentSum := int64(nums[node])
        invertedSum := -currentSum
        
        for _, neighbor := range adjList[node] {
            if neighbor == parent { continue }
            childWithoutInv, childWithInv := dfs(neighbor, node)
            currentSum += max(childWithoutInv, childWithInv)
            invertedSum += max(-childWithoutInv, -childWithInv) // Invert child sums as well.
        }
        return currentSum, invertedSum + 2*int64(nums[node]) // Inverting root node itself.
    }

    // Step 3: Start DFS from root node 0 considering all constraints.
    resultWithoutInversion, resultWithInversion := dfs(0, -1)
    return max(resultWithoutInversion, resultWithInversion) // Return the maximum possible sum.
}
# @lc code=end